### ISSUE TITLE
Also adjust sparsity pattern for DG for nprocs>1

### ISSUE THREAD

**Issue by bangerth** [geodynamics/aspect#1112]
#1100 fixes the sparsity pattern for the DG case when running on one processor. But we are missing the corresponding function in deal.II to also make this happen when using more than one processor. Add it.

**Comment by bangerth** [comment]
When fixing this, we may run into the equivalent of #1113 also for more than one processor, or for other fields.

**Comment by spco** [comment]
@yinghe616 : Are you going to do this? I know we began an implementation of this. If not, I would be interested in doing this within the next few days.

**Comment by yinghe616** [comment]
@spco , I am not working on it currently. If you are interested in fixing it, yes, please go ahead to work on it. If you need a review or discussion later, I'd like to join.

**Comment by spco** [comment]
Ok Ying. Do you have a copy of the code you worked on, that you could email
me? If not I will work from scratch again, no problem.

**Comment by yinghe616** [comment]
@spco , are you planning to work inside the ASPECT? I think perhaps it's better to start the implementation from deal.ii which was also @bangerth 's suggestion I guess. But inside ASPECT, perhaps you can make a non-efficient implementation by just re-setting the continuous elements' interior and face couplings to be zero after the current make_flux_sparsity_parttern. I'll send you a copy of my early code. In case you want to make the function work in deal.ii, I can provide a test case of mixing FEM and DG to test your code.

**Comment by bangerth** [comment]
@yinghe616 had tried to work inside ASPECT, but that required copying a lot of stuff from the deal.II files. In the end, it turned out to be impractical to do it that way, and she got stuck with these basic requirements already. I would suggest to just look at the implementation of the non-DG and DG implementations in deal.II and extend the latter for the functionality of the former. Then put the whole thing into deal.II itself.

**Comment by spco** [comment]
Ok - time to recompile Trilinos, p4est, deal.II devel from source on a Mac... Perhaps this is a good time for me to document that process properly, for up-to-date versions of Xcode, OSX etc...

Once that is done, I guess the workplan should be:
1) Get this function working in deal.II devel
2) Alter ASPECT core.cc to:

```
[usual headers]
#if !DEAL_II_VERSION_GTE(8,5,0)
[extra headers]
#endif
...
[current line 1045:]
    if ((parameters.use_discontinuous_temperature_discretization) || (parameters.use_discontinuous_composition_discretization))
      {
        Table<2,DoFTools::Coupling> face_coupling (introspection.n_components,
                                                   introspection.n_components);
        const typename Introspection<dim>::ComponentIndices &x
          = introspection.component_indices;
        if (parameters.use_discontinuous_temperature_discretization)
          face_coupling[x.temperature][x.temperature] = DoFTools::always;

        if (parameters.use_discontinuous_composition_discretization)
          {
            for (unsigned int c=0; c<parameters.n_compositional_fields; ++c)
              face_coupling[x.compositional_fields[c]][x.compositional_fields[c]] = DoFTools::always;
          }
#if DEAL_II_VERSION_GTE(8,5,0)
        DoFTools::make_flux_sparsity_pattern (dof_handler,
                                              sp,
                                              coupling,
                                              face_coupling,
                                              Utilities::MPI::
                                              this_mpi_process(mpi_communicator));
#else
        if (Utilities::MPI::n_mpi_processes(mpi_communicator) == 1)
          {
            DoFTools::make_sparsity_pattern (dof_handler,
                                             coupling, sp,
                                             constraints, false,
                                             Utilities::MPI::
                                             this_mpi_process(mpi_communicator));
            DoFTools::make_flux_sparsity_pattern (dof_handler,
                                                  sp,
                                                  coupling,
                                                  face_coupling);
          }
        else

          DoFTools::make_flux_sparsity_pattern (dof_handler,
                                                sp,
                                                constraints, false,
                                                Utilities::MPI::
                                                this_mpi_process(mpi_communicator));

      }
    else
      DoFTools::make_sparsity_pattern (dof_handler,
                                       coupling, sp,
                                       constraints, false,
                                       Utilities::MPI::
                                       this_mpi_process(mpi_communicator));
#endif
...
```

What do you think?

**Comment by bangerth** [comment]
Yes. (Though I think your `#ifdef...#endif` straddles an `if`; but general idea is correct.)

For building on Mac, you may want to use the same scripts that are used to build the dmg package.

**Comment by bangerth** [comment]
This can now be addressed given that the deal.II function is in.

**Comment by yinghe616** [comment]
@bangerth  Should I remove "#if DEAL_II_VERSION_GTE(8,5,0)" and all related lines now? Or keep as it is for those early deal.ii.8.4 users?

**Comment by bangerth** [comment]
No, leave these preprocessor defines in for now. We'll take all of them out once we require 8.5.

**Comment by bangerth** [comment]
Oh, I now realize that we *already* require 8.5. So you can take them out from the existing code (in a separate pull request).

**Comment by gassmoeller** [comment]
Was this addressed? Do we now correctly compute the sparsity pattern for DG in parallel?

**Comment by class4kayaker** [comment]
I think this was fixed by #1133.
